rpc/dial.go

```diff
@@ -93,90 +108,161 @@ func dial(
+				// This path is also called by an mdns direct connection and ignores that case.
+				// This will skip all Authenticate/AuthenticateTo calls for the signaler.
+				if !dOpts.usingMDNS && dOpts.authMaterial == "" && dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial != "" {
+					logger.Debug("using signaling's external auth as auth material")
+					dOpts.authMaterial = dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial
+					dOpts.creds = Credentials{}
+				}
 			}
-			fixupWebRTCOptions(dOpts, target, port)
-
-			// When connecting to an external signaler for WebRTC, we assume we can use the external auth's material.
-			// This path is also called by an mdns direct connection and ignores that case.
-			// This will skip all Authenticate/AuthenticateTo calls for the signaler.
-			if !dOpts.usingMDNS && dOpts.authMaterial == "" && dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial != "" {
-				logger.Debug("using signaling's external auth as auth material")
-				dOpts.authMaterial = dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial
-				dOpts.creds = Credentials{}
+
+			if dOpts.debug {
+				logger.Debugw(
+					"trying WebRTC",
+					"signaling_server", dOpts.webrtcOpts.SignalingServerAddress,
+					"host", originalAddress,
+				)
 			}
-		}
 
-		if dOpts.debug {
-			logger.Debugw(
-				"trying WebRTC",
-				"signaling_server", dOpts.webrtcOpts.SignalingServerAddress,
-				"host", originalAddress,
-			)
-		}
+			conn, cached, err := dialFunc(
+				ctxParallel,
+				"webrtc",
+				fmt.Sprintf("%s->%s", dOpts.webrtcOpts.SignalingServerAddress, originalAddress),
+				buildKeyExtra(dOpts),
+				func() (ClientConn, error) {
+					return dialWebRTC(
+						ctxParallel,
+						dOpts.webrtcOpts.SignalingServerAddress,
+						originalAddress,
+						dOpts,
+						logger,
+					)
+				})
+
+			switch {
+			case err == nil:
+				if !cached {
+					logger.Debug("connected via WebRTC")
+				} else if dOpts.debug {
+					logger.Debug("connected via WebRTC (cached)")
+				}
+				dialCh <- dialResult{conn: conn, cached: cached}
+			case !errors.Is(err, ErrNoWebRTCSignaler):
+				// TODO(docs): why don't we try dialing directly after this error?
+				dialCh <- dialResult{err: err, skipDirect: true}
+			case ctxParallel.Err() != nil:
+				dialCh <- dialResult{err: err, skipDirect: true}
+			default:
+				dialCh <- dialResult{err: err}
+			}
+		}()
+	}
 
-		conn, cached, err := dialFunc(
-			ctx,
-			"webrtc",
-			fmt.Sprintf("%s->%s", dOpts.webrtcOpts.SignalingServerAddress, originalAddress),
-			buildKeyExtra(dOpts),
-			func() (ClientConn, error) {
-				return dialWebRTC(
-					ctx,
-					dOpts.webrtcOpts.SignalingServerAddress,
-					originalAddress,
-					dOpts,
-					logger,
-				)
-			})
-		if err == nil {
-			if !cached {
-				logger.Debug("connected via WebRTC")
-			} else if dOpts.debug {
-				logger.Debug("connected via WebRTC (cached)")
+	go func() {
```

Comment:
Took me a bit to understand the function of this goroutine; would you mind adding a comment above it elaborating a bit on your method? I see now that it's closely tied to the `for` loop below. My understanding is that: the `for` loop calls `cancelParallel` to stop the slower method of dialing, the goroutine `Wait`s for the slower method to "finish up" after presumably hitting a `context.Canceled`, and the goroutine closes the `dialCh` to allow the main function to continue running.

Reply:
that's correct - I'll add a comment to explain this flow

Reply:
how's this? https://github.com/viamrobotics/goutils/pull/234/commits/4036d5b7a6a1960f0aec20fc09147161e262f152 - let me know if you'd like me to clarify anything further.

thank you for pushing me to document this better!

Reply:
That comment looks great thank you! Was just getting a little lost in the flow.